f16 vector-vector and matrix-vector afcuda matmul bug

Hello
vector-vector (kind of dotprod) does nt seem to work at half prec but does at f32 :
auto left = af::array(1, 1024, f16); 
auto right = af::array(1024, 1, f16);
af::matmul(left, right) : 

ArrayFire Exception (Function not supported:301):
In function af_err af_matmul(void**, af_array, af_array, af_mat_prop, af_mat_prop)
In file src/api/c/blas.cpp:241
In function af::array af::matmul(const af::array&, const af::array&, af::matProp, af::matProp)
In file src/api/cpp/blas.cpp:18

Note: gevm seems to work (1x1024 matmul 1024x1024).
Note: af::dot() only works for non batched 2d vectors. Would be better for matmul to be consistent between f32 and f16 anyway.

hum, indeed, when trying to instantiate cublasSgemv with half : 
/home/wtambellini/repos/AFWT/src/backend/cuda/blas.cpp: 
In function ‘cuda::gemv_func_def<T> cuda::gemv_func() 
[with T = __half; cuda::gemv_func_def<T> = std::function<cublasStatus_t(cublasContext*, cublasOperation_t, int, int, const __half*, const __half*, int, const __half*, int, const __half*, __half*, int)>]’:

/home/wtambellini/repos/AFWT/src/backend/cuda/blas.cpp:76:16: error: could not convert 
‘cublasSgemv_v2’ from 
‘cublasStatus_t (*)(cublasHandle_t, cublasOperation_t, int, int, const float*, const float*, int, const float*, int, const float*, float*, int) {aka cublasStatus_t (*)(cublasContext*, cublasOperation_t, int, int, const float*, const float*, int, const float*, int, const float*, float*, int)}’ 
to 
‘cuda::gemv_func_def<__half> {aka std::function<cublasStatus_t(cublasContext*, cublasOperation_t, int, int, const __half*, const __half*, int, const __half*, int, const __half*, __half*, int)>}’
         return &cublas##PREFIX##FUNC;           \
                ^
/home/wtambellini/repos/AFWT/src/backend/cuda/blas.cpp:105:1: note: in expansion of macro ‘BLAS_FUNC’
 BLAS_FUNC(gemv, __half, S)  // TODO(umar): Not implemented in CUDA

--- a/src/backend/cuda/blas.cpp
+++ b/src/backend/cuda/blas.cpp
@@ -14,13 +14,17 @@
 #include <platform.hpp>
 
 #include <arith.hpp>
+#include <cast.hpp>
 #include <common/err_common.hpp>
 #include <common/half.hpp>
 #include <complex.hpp>
+#include <copy.hpp>
 #include <cublas.hpp>
 #include <err_cuda.hpp>
 #include <math.hpp>
 #include <reduce.hpp>
+#include <tile.hpp>
+#include <transpose.hpp>
 
 #include <cassert>
 #include <functional>
@@ -57,11 +61,6 @@ using gemmBatched_func_def = std::function<cublasStatus_t(
     const T *, const T **, int, const T **, int, const T *, T **, int, int)>;
 
 template<typename T>
-using gemv_func_def = std::function<cublasStatus_t(
-    cublasHandle_t, cublasOperation_t, int, int, const T *, const T *, int,
-    const T *, int, const T *, T *, int)>;
-
-template<typename T>
 using trsm_func_def = std::function<cublasStatus_t(
     cublasHandle_t, cublasSideMode_t, cublasFillMode_t, cublasOperation_t,
     cublasDiagType_t, int, int, const T *, const T *, int, T *, int)>;
@@ -89,20 +88,6 @@ BLAS_FUNC(gemmBatched, cfloat, C)
 BLAS_FUNC(gemmBatched, double, D)
 BLAS_FUNC(gemmBatched, cdouble, Z)
 BLAS_FUNC(gemmBatched, __half, H)
-
-BLAS_FUNC_DEF(gemv)
-BLAS_FUNC(gemv, float, S)
-BLAS_FUNC(gemv, cfloat, C)
-BLAS_FUNC(gemv, double, D)
-BLAS_FUNC(gemv, cdouble, Z)
-
-template<>
-gemv_func_def<half> gemv_func<half>() {
-    assert(1 != 1 && "GEMV for half is not available.");
-    return gemv_func_def<half>();
-}
-
-// BLAS_FUNC(gemv, __half, S)  // TODO(umar): Not implemented in CUDA
 
 BLAS_FUNC_DEF(trsm)
 BLAS_FUNC(trsm, float, S)
@@ -311,22 +296,9 @@ void gemm(Array<T> &out, af_mat_prop optLhs, af_mat_prop optRhs, const T *alpha,
     dim4 oStrides = out.strides();
 
     if (oDims.ndims() <= 2) {
-        if (rDims[bColDim] == 1) {
-            dim_t incr = (optRhs == AF_MAT_NONE) ? rStrides[0] : rStrides[1];
-            if (is_same<T, half>::value) {
-                AF_ERROR(
-                    "GEMV does not support half, Please create an issue on "
-                    "the GitHub repo.",
-                    AF_ERR_NOT_SUPPORTED);
-            }
-            CUBLAS_CHECK(gemv_func<T>()(blasHandle(), lOpts, lDims[0], lDims[1],
-                                        alpha, lhs.get(), lStrides[1],
-                                        rhs.get(), incr, beta, out.get(), 1));
-        } else {
-            CUBLAS_CHECK(gemmDispatch<T>(blasHandle(), lOpts, rOpts, M, N, K,
-                                         alpha, lhs, lStrides[1], rhs,
-                                         rStrides[1], beta, out, oStrides[1]));
-        }
+        CUBLAS_CHECK(gemmDispatch<T>(blasHandle(), lOpts, rOpts, M, N, K,
+                                     alpha, lhs, lStrides[1], rhs,
+                                     rStrides[1], beta, out, oStrides[1]));
     } else {
         int batchSize = oDims[2] * oDims[3];
         vector<const T *> lptrs(batchSize);
@@ -382,10 +354,9 @@ void gemm(Array<T> &out, af_mat_prop optLhs, af_mat_prop optRhs, const T *alpha,
 template<typename T>
 Array<T> dot(const Array<T> &lhs, const Array<T> &rhs, af_mat_prop optLhs,
              af_mat_prop optRhs) {
-    const Array<T> lhs_ = (optLhs == AF_MAT_NONE ? lhs : conj<T>(lhs));
-    const Array<T> rhs_ = (optRhs == AF_MAT_NONE ? rhs : conj<T>(rhs));
-
-    const Array<T> temp = arithOp<T, af_mul_t>(lhs_, rhs_, lhs_.dims());
+    auto lhs_ = (optLhs == AF_MAT_NONE ? lhs : conj<T>(lhs));
+    auto rhs_ = (optRhs == AF_MAT_NONE ? rhs : conj<T>(rhs));
+    auto temp = arithOp<T, af_mul_t>(lhs_, rhs_, lhs_.dims());
     return reduce<af_add_t, T, T>(temp, 0, false, 0);
 }
 

--- a/test/blas.cpp
+++ b/test/blas.cpp
@@ -31,6 +31,9 @@ using af::max;
 using af::randu;
 using af::setDevice;
 using af::span;
+using af::constant;
+using af::dot;
+using af::transpose;
 using std::copy;
 using std::cout;
 using std::endl;
@@ -328,6 +331,12 @@ TEST(MatrixMultiply, RhsBroadcastBatched) {
 float alpha = 1.f;
 float beta = 0.f;
 
+float h_gold_gemv[4] = {5, 5, 5, 5};
+float h_half_ones[20] = {1.f, 1.f, 1.f, 1.f, 1.f,
+                         1.f, 1.f, 1.f, 1.f, 1.f,
+                         1.f, 1.f, 1.f, 1.f, 1.f,
+                         1.f, 1.f, 1.f, 1.f, 1.f};
+
 float h_lhs[9] =         {1.f, 4.f, 7.f,
                           2.f, 5.f, 8.f,
                           3.f, 6.f, 9.f};
@@ -576,6 +585,11 @@ INSTANTIATE_TEST_CASE_P(
         test_params(AF_MAT_TRANS, AF_MAT_NONE,  &alpha, h_lhs_tall, h_rhs_tall, h_gold_TN, dim4(3, 2), dim4(3, 2), dim4(2, 2), &beta, NULL_ARRAY),
         test_params(AF_MAT_TRANS, AF_MAT_TRANS, &alpha, h_lhs_tall, h_rhs_wide, h_gold_TT, dim4(3, 2), dim4(2, 3), dim4(2, 2), &beta, NULL_ARRAY),
 
+        test_params(AF_MAT_NONE, AF_MAT_NONE, &alpha, h_half_ones, h_half_ones, h_gold_gemv, dim4(4, 5), dim4(5, 1), dim4(4, 1), &beta, NULL_ARRAY),
+        test_params(AF_MAT_NONE, AF_MAT_NONE, &alpha, h_half_ones, h_half_ones, h_gold_gemv, dim4(1, 5), dim4(5, 1), dim4(1, 1), &beta, NULL_ARRAY),
+        test_params(AF_MAT_NONE, AF_MAT_TRANS, &alpha, h_half_ones, h_half_ones, h_gold_gemv, dim4(4, 5), dim4(1, 5), dim4(4, 1), &beta, NULL_ARRAY),
+        test_params(AF_MAT_TRANS, AF_MAT_NONE, &alpha, h_half_ones, h_half_ones, h_gold_gemv, dim4(5, 4), dim4(5, 1), dim4(4, 1), &beta, NULL_ARRAY),
+
         test_params(AF_MAT_NONE,  AF_MAT_NONE,  &alpha, h_lhs_tall, h_rhs_wide, h_gold_NN, dim4(3, 2), dim4(2, 3), dim4(3, 3), &beta, FULL_ARRAY),
         test_params(AF_MAT_NONE,  AF_MAT_TRANS, &alpha, h_lhs_tall, h_rhs_tall, h_gold_NT, dim4(3, 2), dim4(3, 2), dim4(3, 3), &beta, FULL_ARRAY),
         test_params(AF_MAT_TRANS, AF_MAT_NONE,  &alpha, h_lhs_tall, h_rhs_tall, h_gold_TN, dim4(3, 2), dim4(3, 2), dim4(2, 2), &beta, FULL_ARRAY),
@@ -679,4 +693,24 @@ TEST(Gemm, DocSnippet) {
     af_release_array(Csub);
 
     ASSERT_VEC_ARRAY_EQ(gold2, dim4(5, 5, 2), c2);
+}
+
+TEST(Gemv, HalfScalarProduct) {
+    SUPPORTED_TYPE_CHECK(half_float::half);
+
+    const unsigned int sizeValue = 5;
+    array gold = constant(sizeValue, 4, 1, f16);
+    {
+        array a     = constant(1, 4, sizeValue, f16);
+        array b     = constant(1, sizeValue, 1, f16);
+        array mmRes = matmul(a, b);
+        ASSERT_ARRAYS_EQ(mmRes, gold);
+    }
+    {
+        array a     = constant(1, 1, sizeValue, f16);
+        array b     = constant(1, sizeValue, 1, f16);
+        array mmRes = matmul(a, b);
+        array dotRes = dot(transpose(a), b);
+        ASSERT_ARRAYS_EQ(mmRes, dotRes);
+    }
 }

